pkg/volume/util/util.go

```diff
@@ -308,6 +318,115 @@ func LoadPodFromFile(filePath string) (*v1.Pod, error) {
 	return pod, nil
 }
 
+// PopulateZoneConfig populates the zone configurations based on several factors: node.zone; zone/zones from storageclass; allowedTopologies and Replicas
+// For single-zone volumes i.e. replicas = 1:
+//    VolumeZoneConfig.Zone, if present, should be used as the zone of the volume.
+//    VolumeZoneConfig.Zones, if present, should be used to choose a zone for a volume.
+//    VolumeZoneConfig.ReplicaZoneFromNode can be ignored
+//	  if neither Zone/Zones present, query and choose 1 zone from active zones
+// For multi-zone volumes i.e. replicas > 1:
+//    VolumeZoneConfig.ReplicaZoneFromNode, if present, should be used for one of the replica zones. Consider ZonesAlreadyPicked = 1
+//    VolumeZoneConfig.Zones, if present, should be used to choose N zones for the volume where N = replicas - ZonesAlreadyPicked
+//    if VolumeZoneConfig.Zones is absent, query and choose N zones from active zones where N = replicas - ZonesAlreadyPicked
+func PopulateZoneConfig(zonePresent, zonesPresent bool, zone, zones string, node *v1.Node, replicas int, allowedTopologies []v1.TopologySelectorTerm) (*VolumeZoneConfig, error) {
+	var zoneCfg VolumeZoneConfig
+	zoneCfg.ZonePresent = zonePresent
+	zoneCfg.Zone = zone
+	zoneCfg.ZonesPresent = zonesPresent
+	zoneCfg.Zones = zones
+	zoneCfg.ReplicaZoneFromNodePresent = false
+	zoneCfg.ReplicaZoneFromNode = ""
+
+	if zonePresent && zonesPresent {
+		// basic error - fail right away
+		return nil, fmt.Errorf("both zone and zones StorageClass parameters must not be used at the same time")
+	}
+
+	if zonePresent && replicas > 1 {
+		// for multi replica volumes, need zones and not zone
+		return nil, fmt.Errorf("for multi replica volumes, single Zone cannot be specified; require Zones")
+	}
+
+	if !utilfeature.DefaultFeatureGate.Enabled(features.DynamicProvisioningScheduling) {
+		// not much to do - return the passed parameters as is
+		return &zoneCfg, nil
+	}
+
+	allowedZones, err := ZonesFromAllowedTopologies(allowedTopologies)
+	if err != nil {
+		return nil, err
+	}
+	if allowedZones.Len() > 0 {
```

Comment:
Should it return error if `allowedZones.Len() == 0 `? Such `allowedTopologies` are not very useful.

Reply:
Good point. I was not sure if https://github.com/kubernetes/kubernetes/issues/66184 will check for and reject `allowedZones.Len() == 0`. @verult let me know if we need to check for and reject `allowedZones.Len() == 0` here.

Reply:
i'm not sure if api validation could handle this case.

you can have allowed topologies > 0 but not actually specify the zone label